src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFakeImpl.java

```diff
@@ -25,10 +28,33 @@ public void createUrlAlias(UrlAlias urlAlias) {
     return aliases.get(alias);
   }
 
+  /**
+   * Delete the UrlAlias if exists from aliases HashMap.
+   *
+   * @param email of the user to whom the UrlAlias belongs
+   * @param alias the UrlAlias which should be removed
+   *
+   * @throws IllegalArgumentException if no such email or alias
+   */
+
   @Override
   public void deleteUrlAlias(String email, String alias) {
-    // TODO: We should implement it
-    throw new UnsupportedOperationException();
+    UrlAlias foundUrlAlias = null;
+
+    for (UrlAlias urlAlias : aliases.values()) {
+      if (foundUrlAlias == null) {
+        if (urlAlias.email().equals(email) && urlAlias.alias().equals(alias)) {
+          foundUrlAlias = urlAlias;
+          break;
+        }
+      }
+    }
+    if (foundUrlAlias == null) {
+      throw new IllegalArgumentException();
+    }
+    if (aliases.containsKey(alias)) {
```

Comment:
If we can find UrlAlias by the alias, the UrlAlias exists in the hash map. Instead, you need to check if this alias belongs to the user who is trying to delete it.
```suggestion
   if (!alias.email().equals(email)) {
      throw new PermissionDenied();
    }
```

Reply:
I didn't quite understand what do you mean by alias.email
Maybe like this then: !aliases.get(email).equals(email)

Reply:
I changed the function, now it throws all needed exceptions, check it out

Reply:
:+1: